### PR TITLE
nixos/tmp: Make /tmp on ramdisk usable again

### PR DESCRIPTION
**Pull request by dasJ**
@poettering decided we only need a limited number of inodes in our /tmp,
so why not limit that for every systemd user? That makes medium-sized nix
builds impossible so this commit restores the old behaviour which is the
kernel default of half the number of physical RAM pages which does not
seem too unreasonable to me.

###### Motivation for this change

I want to build stuff

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/tmp.nix [hunk 1]
```diff
@@ -30,7 +30,13 @@ with lib;
 
   config = {
 
-    systemd.additionalUpstreamSystemUnits = optional config.boot.tmpOnTmpfs "tmp.mount";
+    systemd.mounts = mkIf config.boot.tmpOnTmpfs [
+      {
+        what = "tmpfs";
+        where = "/tmp";
+        mountConfig.Options = [ "mode=1777" "strictatime" "rw" "nosuid" "nodev" "size=50%" "huge=within_size" ];
```
Comment (flokli) [inline nixos/modules/system/boot/tmp.nix:37]: Can the `huge=within_size` be in a separate commit at least? It feels like a performance optimization that should be done on top of a fix (as in, if it causes regressions, I want to be able to revert it separately).

I'd even argue and say this is something very special that users should probably set on their own (or why is it not a default in general?)

Comment (dasJ) [reply nixos/modules/system/boot/tmp.nix]: I removed the option altogether since people who want to use this functionality can just use the module systems's merge function to add options.

File: nixos/modules/system/boot/tmp.nix [hunk 2]
```diff
@@ -30,7 +30,13 @@ with lib;
 
   config = {
 
-    systemd.additionalUpstreamSystemUnits = optional config.boot.tmpOnTmpfs "tmp.mount";
+    systemd.mounts = mkIf config.boot.tmpOnTmpfs [
+      {
+        what = "tmpfs";
+        where = "/tmp";
+        mountConfig.Options = [ "mode=1777" "strictatime" "rw" "nosuid" "nodev" "size=50%" ];
```
Comment (otavio) [inline nixos/modules/system/boot/tmp.nix:37]: The commit log seems to need to be adjusted as there is no huge page support here.  This seems to have been forgotten to be reworked when changing the commit.

**Review by otavio — CHANGES_REQUESTED** [review]

File: nixos/modules/system/boot/tmp.nix (hunk 2 above)
Comment (flokli) [reply nixos/modules/system/boot/tmp.nix:37]: The commit message doesn't contain any reference to huge page support anymore.

**Comment by cpcloud** [comment]
After applying this change to one of my machines, the `tmp.mount` unit fails to start with:

```
Jan 09 08:44:30 albatross systemd[1]: Mounting /tmp...
Jan 09 08:44:30 albatross mount[7403]: mount: /tmp: special device tmpfs does not exist.
Jan 09 08:44:30 albatross systemd[1]: tmp.mount: Mount process exited, code=exited, status=32/n/a
Jan 09 08:44:30 albatross systemd[1]: tmp.mount: Failed with result 'exit-code'.
Jan 09 08:44:30 albatross systemd[1]: Failed to mount /tmp.
```

This documentation (https://www.freedesktop.org/software/systemd/man/systemd.mount.html) suggests that `Type` isn't necessary, but maybe for some reason it is here?

**Comment by endocrimes** [comment]
@cpcloud hah I just lost too much of my morning to the same - filled a fix in #108860

**Comment by worldofpeace** [comment]
> @poettering decided we only need a limited number of inodes in our /tmp,
so why not limit that for every systemd user? That makes medium-sized nix
builds impossible so this commit restores the old behaviour which is the
kernel default of half the number of physical RAM pages which does not
seem too unreasonable to me.

Let's not be aggressive towards Lennart. We're not the first downstream to have to undo this https://github.com/coreos/fedora-coreos-config/pull/685, and in nixos alone our `systemd` expression undoes a lot of stuff :grin: 
They've since increased it to 50% which is similar to what this PR itself does https://github.com/systemd/systemd/pull/16576.

**Comment by markuskowa** [comment]
Are there any plans to back port this fix to 20.09? I just ran into the same problem, wondering how a build can fail on a 128GB tmpfs:
```
tmpfs on /tmp type tmpfs (rw,nosuid,nodev,nr_inodes=409600)
```
from (20.09.2344.a3a3dda3bac)

**Comment by poscat0x04** [comment]
Is it possible to add a test for this kind of issues? Cause this is a pretty severe issue.

**Comment by dasJ** [comment]
I am really sorry for the problems. My testing was based on 20.09 (which does include `systemd.additionalUpstreamSystemUnits`) so I missed that detail. @ajs124 has a promising fix in #108881 however.

**Comment by vcunat** [comment]
Apparently this doesn't work as expected: #157512  (not sure if something else changed in the meantime)
